### PR TITLE
Prompt Windows UAC when updating in a privileged folder. 

### PR DESCRIPTION
**Pull request by cameronmaske**
Hey there, 
First off, thanks so much for PyUpdater, it's a much-needed package in the python ecosystem and has been immensely useful in my own projects! 

This patch should address an issue in windows, when the user has installed the `.exe` in a privileged folder (e.g. `C://Program Files` aka #35) and the update fails silently. 
It should prompt the user for access. 

I need to do some testing to verify this works on my end and will update this PR with functional proof.

### CODE REVIEW

**Comment by JMSwag** [comment]
Thanks for the PR!

**Comment by JMSwag** [comment]
@cameronmaske I'm working on fixing the CI issues. Once complete please squash your commits & rebase on develop so I can merge. Thanks again for the work.

**Comment by JMSwag** [comment]
@cameronmaske CI fixes are complete.

**Comment by cameronmaske** [comment]
Hey, thanks @JMSwag 
I'm just doing some more functional testing of this on my behalf. Currently, logic that runs command privileged works well, but the code to determine if `.exe` is "protected" has been a touch flaky. I've got a bunch more testing to do (just to ensure the folder logic works) but will rebase post that.

**Comment by JMSwag** [comment]
@cameronmaske NP! 

Sounds good. I'm sure there are quite a few users that will be very happy once this merges.

**Comment by cameronmaske** [comment]
Hmmm, seems this is failing on Python 3 due to `os.popen4`. 
Will try replacing it with https://docs.python.org/3/library/subprocess.html#replacing-os-popen-os-popen2-os-popen3

**Comment by cameronmaske** [comment]
This is really close to working fully.
It works for my application running on Python 2. But the tests appear to fail for Python 3.
I'm a bit at a loss for what to try next currently and don't have much time to invest to get the Python 3 aspect working 😢.
Open to suggestions or if anyone wants to take this further.

**Comment by 2press** [comment]
I am not sure what your problem is, but I tried to run just `def file_require_admin(file_path)` and could make it compatible to Python 2 and 3 by adding a decode command when reading the output:

`output = process.stdout.read().decode("utf-8")`

The problems has been that in Python 3 `process.stdout.read()` is returning a byte-string which is no longer compatible to a normal string. 

But I see **new problems**:

- Python 2 seem to be unable to handle a path or an executable with special chars like ä, ü, ö, while this doesn't seem to be a problem with Python 3. 
-  `return "Access is denied" in output` will only work when Windows is running in English, e. g., my output in German is `Zugriff verweigert; 0 Datei(en) kopiert.` But, it might be sufficient to simply look for `0` or `1` to determine if copying was successful.

File: pyupdater/client/updates.py [hunk 1]
```diff
@@ -49,17 +49,12 @@
...
-    process = subprocess.Popen(
-        cmd,
-        shell=True,
-        stdin=subprocess.PIPE,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT
-    )
-    output = process.stdout.read().decode("utf-8")
-    return "Access is denied" in output
+    try:
```
Comment (cameronmaske) [inline pyupdater/client/updates.py]: Made a change passed on the comment from https://github.com/JMSwag/PyUpdater/pull/116#issuecomment-384422747
Need to test if this operates in the same way....

Comment (2press) [reply pyupdater/client/updates.py]: Works for me on Python 2 and 3 without special chars and on Python 3 with special chars, but not on Python 2 with special chairs. But making the adjustment `with open(file_path.decode('utf-8'), "a"):`  to this, works for me on Python 2 with special chars as well.

Comment (cameronmaske) [reply pyupdater/client/updates.py]: I'll make that change. How are you testing this?

**Comment by 2press** [comment]
I am testing this by copying the `file_require_admin` method into a new small python script and calling it for local files on my system.

File: pyupdater/client/updates.py (hunk 1 above)
Comment (2press) [reply pyupdater/client/updates.py]: I am testing this by copying the `file_require_admin` method into a new small python script and calling it for local files on my system.

**Comment by 2press** [comment]
What's the current problem?

**Comment by cameronmaske** [comment]
@2press I'm not too sure. Looks like https://github.com/JMSwag/PyUpdater/blob/95118c7d00c68dfb7444f62ac6dd9c2f35b8401e/tests/test_pyupdater.py#L357 is failing/timing out

**Comment by 2press** [comment]
It seems to be really close to working fully. Can you pinpoint what you figure the current problem is? And is there a way to debug this problem more closely locally - can one simply run `test_execution_one_dir_restart` on a local machine to reproduce the bug?

**Comment by JMSwag** [comment]
Tests are passing now. Please rebase on master.

**Comment by Berserker66** [comment]
Any news on this? What still needs to be done?

File: tests/test_pyupdater.py [hunk 2]
```diff
@@ -44,6 +44,8 @@
 LOCK_TIMEOUT = 5*60  # ten minutes timeout
 APP_NAME = 'Acme'
 
+import logging
```
Comment (cameronmaske) [inline tests/test_pyupdater.py]: TODO remove this

File: tox.ini [hunk 3]
```diff
@@ -33,7 +33,11 @@ recreate = True
 
 [pytest]
 addopts =
+    -x
     -n 4
+    -s
```
Comment (cameronmaske) [inline tox.ini]: TODO remove these changes

**Comment by JessicaTegner** [comment]
hi @cameronmaske and @JMSwag what is the progress of this pr?

**Comment by cameronmaske** [comment]
@Berserker66 @NicklasMCHD This PR need some more practical functional testing (I only did limited test a while back). In the end, I opted to just install my program in a non-admin directory out of necessity to ship (and it ended up being a non-issue for users). I don't really have the bandwidth currently to fully test this PR. If you are up for it, I'd really encourage you to install try installing PyUpdater from this branch and see if it allows updates in privileged directories (my limited testing a while back seemed to work). Some reports that it works in practice, would help push it along or prompt further changes.

**Comment by JessicaTegner** [comment]
Hi @cameronmaske I would, but I can't even get the test to pass, when I fork and checkout this pr to my local clone.

**Comment by 2press** [comment]
I made pull request to fix some issues, see https://github.com/cameronmaske/PyUpdater/pull/3.

> Fixed the following issues:
>
>- Added required package `pywin32` to setup.py for Windows
>- Removed `decode` from `file_requires_admin` for Python 3
>- Added `requires_admin` function that automatically distinguishes between files and directories (before `file_requires_admin` was sometimes called for directories, yielding false positives).
>- Merged my pull request https://github.com/JMSwag/PyUpdater/pull/142 to fix Key SigningMemoryError and AppVeyor.
>

As far as I am aware there is only one problem left: While running pytest `file_requires_admin` yields a false postive for some executables, e.g. `C:\Users\tasch\AppData\Local\Temp\pytest-of-tasch\pytest-17\popen-gw2\test_execution_one_file_extrac0\update_repo_extract\Acme.exe`; But if I test `file_requires_admin` outside of pytest for the exact same file, it doesn't give a false positive. I would guess that this has something to do with the multiple `gw`'s executing in parallel.

**Comment by 2press** [comment]
I was able to fix the final issue. The currently running executable gives a false positive for `file_requires_admin` as Windows doesn't allow it to be changed while being executed. Therefore the solution is to test it's directory for admin privileges instead. There might be a better solution (?), but this should worker in most user cases. 

For me now all (py)tests pass on Windows 10 with Python 3.7.2 (64-bit).

**Comment by cameronmaske** [comment]
@2press thanks for contributions in getting this across the line! 

> The currently running executable gives a false positive for file_requires_admin as Windows doesn't allow it to be changed while being executed. 

That fits to what I was experiencing too (without realizing it).

**Comment by 2press** [comment]
All builds seem to fail due to an issue of PyInstaller with pip 19, see https://github.com/pyinstaller/pyinstaller/issues/4003 and https://github.com/pypa/pip/issues/6163.

**Comment by 2press** [comment]
My workaround for the pip issue doesn't seem to be compatible with setuptools - @cameronmaske coud you please reverse it? ~~Maybe needs to be just `pyinstaller >= 3.0 --no-use-pep517` otherwise~~ I would suggest to either pin pip to 18.1 or wait for a update of pip, which seems to be imminent.

**Comment by 2press** [comment]
I have a pull request ready that should push this one over the line: https://github.com/cameronmaske/PyUpdater/pull/5.

**Comment by 2press** [comment]
@JMSwag This pull request should now be ready to be merged.

**Comment by JessicaTegner** [comment]
he @JMSwag  what's the progress on this pr?

**Comment by 2press** [comment]
Quick update from me: I have used this branch in production for one month now and everything is working as desired.

**Comment by JMSwag** [comment]
@2press thanks for the update

**Comment by cameronmaske** [comment]
Hurra! Thanks for getting this across the line @2press

**Pull request merged**